pkg/serviceaccount/claims.go

```diff
@@ -77,7 +84,16 @@ func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirati
 			UID:  string(secret.UID),
 		}
 	}
-	return sc, pc
+
+	// Tweak expiry fields for safe transition of projected service account token.
+	// Warn (instead of fail) after requested expiration time.
+	// Fail after hard-coded extended expiration time.
+	if extendExpiry && expirationSeconds == expirationSecondsSpecial {
+		sc.Expiry = jwt.NewNumericDate(now.Add(time.Duration(expirationExtensionHours) * time.Hour))
```

Comment:
I'd expect this to be bounded to the lesser of `--service-account-max-token-expiration` or 1 year... doing the computation in TokenREST#Create as suggested in https://github.com/kubernetes/kubernetes/pull/89549/files#r399304171 would give you access to the maxExpirationSeconds

Reply:
I wonder what people usually set for this flag.
GKE kind of set this to 48h, which makes me feel like it might not be a good idea setting our extended expiration capped by this.

Reply:
The purpose of the max-expiration flag is to let the admin know when it is safe to rotate the signing key. If we use that key to issue tokens valid for longer than the max expiration, it no longer serves that purpose. I would also accept erroring on startup if extend-expiration was true and max-expiration was less than one year

Reply:
I would be leaning towards erroring if extend-expiration was enabled but max-expiration was set to a short time, instead of silently make the token expire in a shorter time when admin was not aware these two flags are related.

Reply:
that's fine with me

Reply:
Done.